### PR TITLE
added namespace support; new client for v2 api

### PR DESCRIPTION
**Pull request by ikasarov**
LMCROSSITXSADEPLOY-1571

also added script to regen the client

### CODE REVIEW

File: bin/regen-client.sh [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+check_dir "${regen_folder_v2}"
+
+echo -e "Assuming controller project is under this parent dir: ${GREEN}${git_home}${BORING}"
+echo -e "Assuming plugin project is under this parent dir: ${GREEN}${go_home}${BORING}"
+echo -e "Reading from\n\t${GREEN}${definition_file}\n\t${definition_file_v2}${BORING}\nGenerating in \n\t${GREEN}${regen_folder}\n\t${regen_folder_v2}${BORING}"
+
+cd "${git_home%/*}"
+mvn -f "${git_home}/multiapps-controller/" clean package -DskipTests=true -pl=com.sap.cloud.lm.sl.cf.api
+
+swagger generate client -f ${definition_file} -A http_mta -c ${client_name} -t ${regen_folder}
```
Comment (theghost5800) [inline bin/regen-client.sh:53]: It should be downloaded automatically swagger generator if the user doesn't have it installed or at least to report to user to install manually swagger generator with some steps

File: clients/models/metadata.go [hunk 2]
```diff
@@ -30,11 +33,6 @@ type Metadata struct {
 
 // Validate validates this metadata
 func (m *Metadata) Validate(formats strfmt.Registry) error {
-	var res []error
-
-	if len(res) > 0 {
```
Comment (theghost5800) [inline clients/models/metadata.go:35]: Why this validation was removed?

File: clients/mtaclient_v2/mta_v2_client_operations.go [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package mtaclient_v2
+
+import (
+	models "github.com/cloudfoundry-incubator/multiapps-cli-plugin/clients/models"
+	strfmt "github.com/go-openapi/strfmt"
+)
+
+// MtaClientOperations drun drun drun x2
```
Comment (theghost5800) [inline clients/mtaclient_v2/mta_v2_client_operations.go:8]: If this is not auto generated comment, it should be deleted

File: commands/blue_green_deploy_command.go [hunk 4]
```diff
@@ -32,17 +32,15 @@ func (c *BlueGreenDeployCommand) GetPluginCommand() plugin.Command {
...
-				util.GetShortOption(deleteServicesOpt):             "Recreate changed services / delete discontinued services",
+				extDescriptorsOpt:                      "Extension descriptors",
+				deployServiceURLOpt:                    "Deploy service URL, by default 'deploy-service.<system-domain>'",
+				timeoutOpt:                             "Start timeout in seconds",
+				versionRuleOpt:                         "Version rule (HIGHER, SAME_HIGHER, ALL)",
+				operationIDOpt:                         "Active deploy operation ID",
+				actionOpt:                              "Action to perform on active deploy operation (abort, retry, monitor)",
+				forceOpt:                               "Force deploy without confirmation for aborting conflicting processes",
+				util.GetShortOption(noStartOpt):        "Do not start apps",
+				util.GetShortOption(deleteServicesOpt): "Recreate changed services / delete discontinued services",
```
Comment (theghost5800) [inline commands/blue_green_deploy_command.go:43]: Git shows that options related with namespace were deleted

File: commands/deploy_command.go [hunk 5]
```diff
@@ -19,12 +19,12 @@ import (
...
-	versionRuleOpt             = "version-rule"
-	noStartOpt                 = "no-start"
-	useNamespacesOpt           = "use-namespaces"
-	noNamespacesForServicesOpt = "no-namespaces-for-services"
+	extDescriptorsOpt = "e"
+	timeoutOpt        = "t"
+	versionRuleOpt    = "version-rule"
+	noStartOpt        = "no-start"
+	//useNamespacesOpt           = "use-namespaces"
+	//noNamespacesForServicesOpt = "no-namespaces-for-services"
```
Comment (theghost5800) [inline commands/deploy_command.go:27]: Why namespace command line options are commented?

File: commands/deploy_command.go [hunk 6]
```diff
@@ -268,8 +270,10 @@ func (c *DeployCommand) Execute(args []string) ExecutionStatus {
 		return Failure
 	}
 
+	ui.Say("Command line value for namespace: %s\n", terminal.EntityNameColor(namespace))
```
Comment (theghost5800) [inline commands/deploy_command.go:273]: Is this a debug message?

File: commands/file_uploader.go [hunk 7]
```diff
@@ -39,7 +41,7 @@ func (f *FileUploader) UploadFiles() ([]*models.FileMetadata, ExecutionStatus) {
 	log.Tracef("Uploading files '%v'\n", f.files)
 
 	// Get all files that are already uploaded
-	uploadedMtaFiles, err := f.mtaClient.GetMtaFiles()
+	uploadedMtaFiles, err := f.mtaClient.GetMtaFiles(&f.namespace)
```
Comment (theghost5800) [inline commands/file_uploader.go:44]: rename FileUpload variable with more descriptive name

File: commands/mta_command.go [hunk 8]
```diff
@@ -48,13 +49,17 @@ func (c *MtaCommand) Execute(args []string) ExecutionStatus {
...
+	flags.StringVar(&namespace, namespaceOpt, "", "")
+	// namespace := *flags.String(namespaceOpt, "", "")
 	parser := NewCommandFlagsParser(flags, NewDefaultCommandFlagsParser([]string{"MTA_ID"}), NewDefaultCommandFlagsValidator(map[string]bool{}))
 	err = parser.Parse(args)
 	if err != nil {
 		c.Usage(err.Error())
 		return Failure
 	}
 	mtaID := args[0]
+	ui.Say("Command line value for namespace: %s\n", terminal.EntityNameColor(namespace))
```
Comment (theghost5800) [inline commands/mta_command.go:62]: Is this a debug message?

File: commands/mta_command.go [hunk 9]
```diff
@@ -86,10 +91,16 @@ func (c *MtaCommand) Execute(args []string) ExecutionStatus {
 		return Failure
 
 	}
+	if len(mtas) != 1 {
```
Comment (theghost5800) [inline commands/mta_command.go:94]: I think that it will be better to use `len(mtas) > 1`

File: commands/mtas_command.go [hunk 10]
```diff
@@ -58,14 +58,14 @@ func (c *MtasCommand) Execute(args []string) ExecutionStatus {
...
-	mtaClient, err := c.NewMtaClient(host)
+	mtaV2Client, err := c.NewMtaV2Client(host)
 	if err != nil {
 		ui.Failed("Could not get space ID: %s", baseclient.NewClientError(err))
 		return Failure
 	}
 
 	// Get all deployed components
-	mtas, err := mtaClient.GetMtas()
+	mtas, err := mtaV2Client.GetMtasForThisSpace(nil, nil)
```
Comment (theghost5800) [inline commands/mtas_command.go:68]: Is it possible to add function without parameters?

File: commands/undeploy_command.go [hunk 11]
```diff
@@ -33,22 +33,24 @@ func (c *UndeployCommand) GetPluginCommand() plugin.Command {
...
+				actionOpt:                              "Action to perform on the active undeploy operation (abort, retry, monitor)",
+				forceOpt:                               "Force undeploy without confirmation",
+				util.GetShortOption(deleteServicesOpt): "Delete services",
 				util.GetShortOption(deleteServiceKeysOpt):          "Delete existing service keys",
 				util.GetShortOption(deleteServiceBrokersOpt):       "Delete service brokers",
 				util.GetShortOption(noRestartSubscribedAppsOpt):    "Do not restart subscribed apps, updated during the undeployment",
 				util.GetShortOption(noFailOnMissingPermissionsOpt): "Do not fail on missing permissions for admin operations",
 				util.GetShortOption(abortOnErrorOpt):               "Auto-abort the process on any errors",
 				util.GetShortOption(retriesOpt):                    "Retry the operation N times in case a non-content error occurs (default 3)",
+				// TODO: find out what's going on here??
```
Comment (theghost5800) [inline commands/undeploy_command.go:52]: What is the issue here?

File: commands/deploy_command.go [hunk 12]
```diff
@@ -81,24 +81,23 @@ func (c *DeployCommand) GetPluginCommand() plugin.Command {
...
+				deployServiceURLOpt:                    "Deploy service URL, by default 'deploy-service.<system-domain>'",
+				timeoutOpt:                             "Start timeout in seconds",
+				versionRuleOpt:                         "Version rule (HIGHER, SAME_HIGHER, ALL)",
+				operationIDOpt:                         "Active deploy operation ID",
+				actionOpt:                              "Action to perform on active deploy operation (abort, retry, monitor)",
+				forceOpt:                               "Force deploy without confirmation for aborting conflicting processes",
+				moduleOpt:                              "Deploy list of modules which are contained in the deployment descriptor, in the current location",
+				resourceOpt:                            "Deploy list of resources which are contained in the deployment descriptor, in the current location",
+				util.GetShortOption(noStartOpt):        "Do not start apps",
+				util.GetShortOption(namespaceOpt):      "Namespace for the mta, applied to app and service names as well",
```
Comment (boyan-velinov) [inline commands/deploy_command.go:99]: Mark it as EXPERIMENTAL

File: clients/mtaclient/mta_client_operations.go [hunk 13]
```diff
@@ -11,14 +11,14 @@ import (
 type MtaClientOperations interface {
 	ExecuteAction(operationID, actionID string) (ResponseHeader, error)
 	GetMta(mtaID string) (*models.Mta, error)
-	GetMtaFiles() ([]*models.FileMetadata, error)
+	GetMtaFiles(namespace *string) ([]*models.FileMetadata, error)
```
Comment (enchobelezirev) [inline clients/mtaclient/mta_client_operations.go:14]: In Go, and in C++, it is a practise(not that good) that you can provide pointer to an object and later on change the value of the pointer in the calling function. This could be useful in some cases but I think that in your case it would be better to pass the variable by value. I suppose you do not want to change it in the method, right? :)

File: clients/mtaclient/mta_rest_client.go [hunk 14]
```diff
@@ -65,9 +66,10 @@ func (c MtaRestClient) GetMta(mtaID string) (*models.Mta, error) {
 	return result.(*operations.GetMtaOK).Payload, nil
 }
 
-func (c MtaRestClient) GetMtaFiles() ([]*models.FileMetadata, error) {
+func (c MtaRestClient) GetMtaFiles(namespace *string) ([]*models.FileMetadata, error) {
 	params := &operations.GetMtaFilesParams{
-		Context: context.TODO(),
+		Context:   context.TODO(),
+		Namespace: namespace,
```
Comment (enchobelezirev) [inline clients/mtaclient/mta_rest_client.go:72]: If you decide to go without the pointer, this code line would look like this:
Namespace: &namespace

File: clients/mtaclient/operations/upload_mta_file_parameters.go [hunk 15]
```diff
@@ -56,6 +56,12 @@ for the upload mta file operation typically these are written to a http.Request
 */
 type UploadMtaFileParams struct {
 
+	/*Namespace
+	  file namespace
+
+	*/
+	Namespace *string
```
Comment (enchobelezirev) [inline clients/mtaclient/operations/upload_mta_file_parameters.go:63]: We are assigning the files to the namespaces?

File: clients/mtaclient_v2/mta_v2_client_operations.go (hunk 3 above)
Comment (enchobelezirev) [reply clients/mtaclient_v2/mta_v2_client_operations.go:8]: Do not delete it - fix it :D

File: clients/mtaclient_v2/mta_v2_client_operations.go [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+type MtaV2ClientOperations interface {
+	GetMtas(name, namespace *string, spaceGuid string) ([]*models.Mta, error)
+	GetMtasForThisSpace(name, namespace *string) ([]*models.Mta, error)
```
Comment (enchobelezirev) [inline clients/mtaclient_v2/mta_v2_client_operations.go:11]: Would it be better to be like this:
`GetMtasForNamespace(name, namespace *string) ([]*models.Mta, error)`
Something further, you could try to do some refactoring here which would be to introduce a Builder for the request parameters. Meaning that the code will become:
`GetMtas(MtasRequestParameters parameters) ([]*models.Mta, error)`

File: commands/deploy_command.go [hunk 17]
```diff
@@ -183,6 +179,11 @@ func GetStringOpt(name string, optionValues map[string]interface{}) string {
 	return *optionValues[name].(*string)
 }
 
+// GetNullableStringOpt gets the pointer identified by the specified name.
+func GetNullableStringOpt(name string, optionValues map[string]interface{}) *string {
```
Comment (enchobelezirev) [inline commands/deploy_command.go:183]: Why is this needed?

File: commands/mta_command.go [hunk 18]
```diff
@@ -68,14 +73,14 @@ func (c *MtaCommand) Execute(args []string) ExecutionStatus {
 		terminal.EntityNameColor(context.Space), terminal.EntityNameColor(context.Username))
 
 	// Create new REST client
-	mtaClient, err := c.NewMtaClient(host)
+	mtaV2Client, err := c.NewMtaV2Client(host)
```
Comment (enchobelezirev) [inline commands/mta_command.go:76]: I do not remember but have we targeted some backwards compatibility?

File: commands/mtas_command.go (hunk 10 above)
Comment (enchobelezirev) [reply commands/mtas_command.go:68]: Checkout my comment for the refactoring

File: util/archive_handler.go [hunk 19]
```diff
@@ -15,18 +15,20 @@ import (
...
-// GetMtaIDFromArchive retrieves MTA ID from MTA archive
-func GetMtaIDFromArchive(mtaArchiveFilePath string) (string, error) {
+// GetMtaDescriptorFromArchive retrieves MTA ID from MTA archive
+func GetMtaDescriptorFromArchive(mtaArchiveFilePath string) (MtaDescriptor, error) {
+	var descriptor MtaDescriptor
 	// Open the mta archive
 	mtaArchiveReader, err := zip.OpenReader(mtaArchiveFilePath)
 	if err != nil {
-		return "", err
+		return descriptor, err
```
Comment (enchobelezirev) [inline util/archive_handler.go:31]: change to return MtaDescriptor{}

File: clients/models/metadata.go (hunk 2 above)
Comment (ikasarov) [reply clients/models/metadata.go:35]: it's an autogenerated var that checks nothing since in this case there's nothing to validate.

File: clients/mtaclient/mta_client_operations.go (hunk 13 above)
Comment (ikasarov) [reply clients/mtaclient/mta_client_operations.go:14]: This was done since the namespace is an optional parameter and can be an empty reference. A similar thing is done below for 'GetMtaOperations(mtaId *string...'

File: commands/mta_command.go (hunk 18 above)
Comment (ikasarov) [reply commands/mta_command.go:76]: the v2 backend is in itself backwards compatible (or should be, at least)
